perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -4980,17 +5039,23 @@ public MemberGroupStatus getTotalMemberGroupStatus(PerunSession session, Member
 		return groupsManagerImpl.getTotalMemberGroupStatus(session, member, group);
 	}
 
+	@Override
+	public Map<Member, MemberGroupStatus> getTotalGroupStatusForMembers(PerunSession session, Group group, List<Member> members) {
+		return groupsManagerImpl.getTotalGroupStatusForMembers(session, group, members);
+	}
+
 	/**
 	 * Calculates the state of given member in given group and if
 	 * it differs from given 'previousState' calls this method recursively
 	 * for all parent groups.
 	 *
 	 * @param member member
 	 * @param group group
+	 * @param previousStatus previousStatus
 	 * @throws InternalErrorException internal error
 	 */
 	@Override
-	public void recalculateMemberGroupStatusRecursively(PerunSession sess, Member member, Group group) {
+	public void recalculateMemberGroupStatusRecursively(PerunSession sess, Member member, Group group, Map<Pair<Group, Member>, MemberGroupStatus> previousStatus) {
```

Comment:
When I'm thinking about this again, I think it might be better to use 
```Java
Map<Integer, Map<Integer, MemberGroupStatus>>
```
to store the statuses (where the first integer is groupId and the second one is memberId).
The problem that I worry about is with our inheritance. Someone could place a `RichGroup` or a `RichMember` to the map together with `Group` or `Map`. That would lead to inconsistent behaviour of the `hashcode()` and `equals()` methods that might cause some bugs that are very hard to find and debug.

Reply:
I haven't thought and I think the Map looks better, changed, thanks.